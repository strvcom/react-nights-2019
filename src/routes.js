import { generatePath } from 'react-router'

export const HOMEPAGE = '/'
export const PRODUCT_LIST = '/products'
export const PRODUCT_DETAIL = '/products/:productId'
export const getProductDetailRoute = productId =>
  generatePath(PRODUCT_DETAIL, { productId })
export const CART = '/cart'
export const SIGN_UP = '/signup'
export const LOGIN = '/login'
export const ACCOUNT = '/account'
export const LOGOUT = '/logout'
