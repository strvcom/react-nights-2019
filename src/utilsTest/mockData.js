const FIRST_PRODUCT = {
  id: 1,
  attributes: {
    name: 'First Test Product',
    description:
      'Lorem ipsum dolor sit amet, consectetur adipiscing elit. Etiam pellentesque in neque vitae tincidunt. In gravida eu ipsum non condimentum. Curabitur libero leo, gravida a dictum vestibulum, sollicitudin vel quam.',
    image_url: 'https://img.commercelayer.io/skus/1.png',
  },
  relationships: {
    prices: {
      data: [{ id: 1 }],
    },
  },
}

const SECOND_PRODUCT = {
  id: 2,
  attributes: {
    name: 'Second Test Product',
    description:
      'Lorem ipsum dolor sit amet, consectetur adipiscing elit. Etiam pellentesque in neque vitae tincidunt. In gravida eu ipsum non condimentum. Curabitur libero leo, gravida a dictum vestibulum, sollicitudin vel quam.',
    image_url: 'https://img.commercelayer.io/skus/1.png',
  },
  relationships: {
    prices: {
      data: [{ id: 2 }],
    },
  },
}

const INCLUDES = [
  {
    id: 1,
    attributes: {
      amount_cents: 1000,
      amount_float: 10,
      formatted_amount: '$10',
    },
  },
  {
    id: 2,
    attributes: {
      amount_cents: 2000,
      amount_float: 20,
      formatted_amount: '$20',
    },
  },
]

export const PRODUCTS = {
  data: [FIRST_PRODUCT, SECOND_PRODUCT],
  included: INCLUDES,
  meta: {
    page_count: 1,
    record_count: 2,
  },
}

export const PRODUCT = {
  data: FIRST_PRODUCT,
  included: INCLUDES,
}

export const USER = {
  id: 1,
  email: 'michal@test.com',
  firstName: 'Michal',
  password: 'Heslo1234',
}

export const TOKEN = {
  access_token: 'accessToken',
  token_type: 'bearer',
  expires_in: 7200,
  refresh_token: 'refreshToken',
  scope: 'market:335',
  created_at: 1555858950,
  owner_id: 1,
  owner_type: 'customer',
}

export const LOGIN_DATA = {
  id: USER.id,
  type: 'customers',
  links: {
    self: `https://the-amber-brand-12.commercelayer.io/api/customers/${
      USER.id
    }`,
  },
  attributes: {
    email: USER.email,
    status: 'prospect',
    created_at: '2019-04-21T15:02:29.631Z',
    updated_at: '2019-04-21T15:02:29.631Z',
    reference: null,
    metadata: { firstName: USER.firstName },
  },
}

export const ERROR_EMAIL_ALREADY_TAKEN = {
  title: 'has already been taken',
  detail: 'email - has already been taken',
  code: 'VALIDATION_ERROR',
  source: { pointer: '/data/attributes/email' },
  status: '422',
  meta: {
    error: 'taken',
    value: 'test@test.com',
  },
}
