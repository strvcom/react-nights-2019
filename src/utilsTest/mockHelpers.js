import fetchMock from 'fetch-mock'
import config from '../config'
import {
  PRODUCTS,
  PRODUCT,
  USER,
  LOGIN_DATA,
  TOKEN,
  ERROR_EMAIL_ALREADY_TAKEN,
} from './mockData'

export const mockUnauthorizedRequest = () =>
  fetchMock.restore().post(`${config.apiUrl}/oauth/token`, {
    status: 401,
    body: {},
  })

export const mockAuthorizedRequest = () =>
  fetchMock
    .restore()
    .post(`${config.apiUrl}/oauth/token`, {
      status: 200,
      body: TOKEN,
    })
    .get(`${config.apiUrl}/api/customers/${USER.id}`, {
      status: 200,
      body: {
        data: LOGIN_DATA,
      },
    })

export const mockFetchProducts = () =>
  mockUnauthorizedRequest().get(
    `${
      config.apiUrl
    }/api/skus?include=prices&page%5Bnumber%5D=1&page%5Bsize%5D=50`,
    {
      status: 200,
      body: PRODUCTS,
    }
  )

export const mockFetchProduct = () =>
  mockUnauthorizedRequest().get(`${config.apiUrl}/api/skus/1?include=prices`, {
    status: 200,
    body: PRODUCT,
  })

export const mockSignUp422 = () =>
  mockUnauthorizedRequest().post(`${config.apiUrl}/api/customers`, {
    status: 411,
    body: {
      errors: [ERROR_EMAIL_ALREADY_TAKEN],
    },
  })

export const mockSignUp201 = () =>
  mockAuthorizedRequest().post(`${config.apiUrl}/api/customers`, {
    status: 201,
    body: {
      data: LOGIN_DATA,
    },
  })
