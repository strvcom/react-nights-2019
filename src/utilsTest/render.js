import React from 'react'
import { MemoryRouter } from 'react-router-dom'
import { render } from 'react-testing-library'

export const renderWithRouter = (component, uri) =>
  render(<MemoryRouter initialEntries={[uri]}>{component}</MemoryRouter>)
