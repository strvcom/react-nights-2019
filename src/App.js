import React from 'react'
import { Switch, Route, Redirect } from 'react-router-dom'
import { Provider } from 'react-redux'
import { ToastContainer, toast } from 'react-toastify'

import GlobalStyles from './globalStyles'
import { ProductList } from './pages/ProductList'
import { ProductDetail } from './pages/ProductDetail'
import { Cart } from './pages/Cart'
import { SignUp } from './pages/SignUp'
import { LogIn } from './pages/LogIn'
import { Logout } from './pages/Logout'
import { Account } from './pages/Account'
import { NotFound } from './pages/NotFound'
import { PrivateRoute } from './components/PrivateRoute'
import { ErrorBoundary } from './components/ErrorBoundary'
import { getCustomer } from './utils/customer'
import { configureStore } from './store'
import * as routes from './routes'

const defaultStore = configureStore({
  customer: getCustomer(),
})

const App = ({ store }) => (
  <Provider store={store || defaultStore}>
    <React.Fragment>
      <GlobalStyles />
      <ToastContainer position={toast.POSITION.BOTTOM_RIGHT} />
      <ErrorBoundary>
        <Switch>
          <Route
            path={routes.HOMEPAGE}
            exact
            render={() => <Redirect to={routes.PRODUCT_LIST} />}
          />
          <Route path={routes.PRODUCT_LIST} exact component={ProductList} />
          <Route path={routes.PRODUCT_DETAIL} component={ProductDetail} />
          <Route path={routes.CART} component={Cart} />
          <Route path={routes.SIGN_UP} component={SignUp} />
          <Route path={routes.LOGIN} component={LogIn} />
          <Route path={routes.LOGOUT} component={Logout} />
          <PrivateRoute path={routes.ACCOUNT} component={Account} />
          <Route component={NotFound} />
        </Switch>
      </ErrorBoundary>
    </React.Fragment>
  </Provider>
)

export { App }
