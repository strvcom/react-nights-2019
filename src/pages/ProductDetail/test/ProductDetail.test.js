import React from 'react'
import { waitForElement } from 'react-testing-library'
import 'jest-styled-components'

import { App } from '../../../App'
import * as routes from '../../../routes'
import { renderWithRouter } from '../../../utilsTest/render'
import { mockFetchProduct } from '../../../utilsTest/mockHelpers'

describe('[pages] ProductDetail', () => {
  const productId = 1

  describe('when loading', () => {
    it('should render correctly', () => {
      const renderer = renderWithRouter(
        <App />,
        routes.getProductDetailRoute(productId)
      )
      expect(renderer.container).toMatchSnapshot()
    })
  })

  describe('when product loaded', () => {
    mockFetchProduct()

    it('should render correctly', async () => {
      const renderer = renderWithRouter(
        <App />,
        routes.getProductDetailRoute(productId)
      )
      await waitForElement(() => renderer.getByTestId('product-detail'))
      expect(renderer.container).toMatchSnapshot()
    })
  })
})
