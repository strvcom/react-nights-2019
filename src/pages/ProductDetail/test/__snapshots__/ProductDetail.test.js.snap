// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[pages] ProductDetail when loading should render correctly 1`] = `
.c2 {
  padding: 2rem;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  border-bottom: 0.1rem solid gainsboro;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  padding: 3rem;
}

.c1 {
  margin: 0 1rem;
}

.c4 {
  width: 100%;
  padding: 5rem 0;
  padding: 5rem 0;
}

.c5 {
  width: 5rem;
  height: 5rem;
  margin: 0 auto;
}

.c6 {
  -webkit-animation: cZxgpV 2s linear infinite;
  animation: cZxgpV 2s linear infinite;
  height: 100%;
  -webkit-transform-origin: center center;
  -ms-transform-origin: center center;
  transform-origin: center center;
  width: 100%;
}

.c7 {
  stroke: #ef0d33;
  stroke-dasharray: 1,200;
  stroke-dashoffset: 0;
  -webkit-animation: iwIXhf 1.5s ease-in-out infinite;
  animation: iwIXhf 1.5s ease-in-out infinite;
  stroke-linecap: round;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

<div>
  <div
    class="Toastify"
  />
  <header
    class="c0"
  >
    <div
      class=""
    >
      <a
        class="c1"
        href="/products"
      >
        All Products
      </a>
    </div>
    <div
      class=""
    >
      <a
        class="c1"
        href="/cart"
      >
        My Cart
      </a>
      |
      <a
        class="c1"
        href="/login"
      >
        Log In
      </a>
       |
      <a
        class="c1"
        href="/signup"
      >
        Sign Up
      </a>
    </div>
  </header>
  <div
    class="c2"
  >
    <div
      class="c3"
    >
      <div
        class="c4"
        data-testid="loader"
      >
        <div
          class="c5"
        >
          <svg
            class="c6"
            viewBox="25 25 50 50"
          >
            <circle
              class="c7"
              cx="50"
              cy="50"
              fill="none"
              r="20"
              stroke-miterlimit="10"
              stroke-width="4"
            />
          </svg>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`[pages] ProductDetail when product loaded should render correctly 1`] = `
.c2 {
  padding: 2rem;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  border-bottom: 0.1rem solid gainsboro;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  padding: 3rem;
}

.c1 {
  margin: 0 1rem;
}

.c7 {
  font-family: circular,serif;
  font-weight: 400;
  color: #ef0d33;
  text-transform: uppercase;
  text-align: center;
}

.c10 {
  background: #ef0d33;
  cursor: pointer;
  padding: 1rem;
  margin-top: 0.5rem;
  border: none;
  border-radius: 5px;
  color: #fff;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.c4 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background-color: white;
}

.c5 {
  height: 400px;
}

.c6 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  max-width: 40%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background-color: #f9f9f9;
  padding: 0 3rem;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c8 {
  font-size: 2.4rem;
  color: #ef0d33;
}

.c9 {
  margin: 2rem 0;
  text-align: center;
}

<div>
  <div
    class="Toastify"
  />
  <header
    class="c0"
  >
    <div
      class=""
    >
      <a
        class="c1"
        href="/products"
      >
        All Products
      </a>
    </div>
    <div
      class=""
    >
      <a
        class="c1"
        href="/cart"
      >
        My Cart
      </a>
      |
      <a
        class="c1"
        href="/login"
      >
        Log In
      </a>
       |
      <a
        class="c1"
        href="/signup"
      >
        Sign Up
      </a>
    </div>
  </header>
  <div
    class="c2"
  >
    <div
      class="c3"
    >
      <div
        data-testid="product-detail"
      >
        <div
          class="c4"
        >
          <img
            class="c5"
            src="https://img.commercelayer.io/skus/1.png"
          />
        </div>
        <div
          class="c6"
        >
          <h1
            class="c7"
          >
            First Test Product
          </h1>
          <span
            class="c8"
          >
            $10
          </span>
          <span
            class="c9"
          >
            Lorem ipsum dolor sit amet, consectetur adipiscing elit. Etiam pellentesque in neque vitae tincidunt. In gravida eu ipsum non condimentum. Curabitur libero leo, gravida a dictum vestibulum, sollicitudin vel quam.
          </span>
          <button
            class="c10"
          >
            Add to Cart
          </button>
          <a
            href="/products"
          >
            Back
          </a>
        </div>
      </div>
    </div>
  </div>
</div>
`;
